src/main/java/com/google/devtools/build/lib/bazel/bzlmod/BazelModuleResolutionFunction.java

```diff
@@ -0,0 +1,188 @@
+      return;
+    }
+    boolean failure = false;
+    for (Map.Entry<String, ModuleKey> dep : discoveredRootModule.getDeps().entrySet()) {
+      ModuleKey resolved = resolvedRootModule.getDeps().get(dep.getKey());
+      if (!dep.getValue().equals(resolved)) {
+        String message =
+            String.format(
+                "For repository '%s', the root module requires module version %s, but got %s in the"
+                    + " resolved dependency graph.",
+                dep.getKey(), dep.getValue(), resolved);
+        if (mode == CheckDirectDepsMode.WARNING) {
+          env.getListener().handle(Event.warn(message));
+        } else {
+          env.getListener().handle(Event.error(message));
+          failure = true;
+        }
+      }
+    }
+    if (failure) {
+      throw new BazelModuleResolutionFunctionException(
+          ExternalDepsException.withMessage(
+              Code.VERSION_RESOLUTION_ERROR, "Direct dependency check failed."),
+          Transience.PERSISTENT);
+    }
+  }
+
+  @VisibleForTesting
+  static BazelModuleResolutionValue createValue(
+      ImmutableMap<ModuleKey, Module> depGraph, ImmutableMap<String, ModuleOverride> overrides)
+      throws BazelModuleResolutionFunctionException {
+    // Build some reverse lookups for later use.
+    ImmutableMap<String, ModuleKey> canonicalRepoNameLookup =
+        depGraph.keySet().stream()
+            .collect(toImmutableMap(ModuleKey::getCanonicalRepoName, key -> key));
+    ImmutableMap<String, ModuleKey> moduleNameLookup =
+        depGraph.keySet().stream()
+            // The root module is not meaningfully used by this lookup so we skip it (it's
+            // guaranteed to be the first in iteration order).
+            .skip(1)
+            .filter(key -> !(overrides.get(key.getName()) instanceof MultipleVersionOverride))
+            .collect(toImmutableMap(ModuleKey::getName, key -> key));
+
+    // For each extension usage, we resolve (i.e. canonicalize) its bzl file label. Then we can
+    // group all usages by the label + name (the ModuleExtensionId).
+    ImmutableTable.Builder<ModuleExtensionId, ModuleKey, ModuleExtensionUsage>
+        extensionUsagesTableBuilder = ImmutableTable.builder();
+    for (Module module : depGraph.values()) {
+      LabelConversionContext labelConversionContext =
+          new LabelConversionContext(
+              StarlarkBazelModule.createModuleRootLabel(module.getCanonicalRepoName()),
+              module.getRepoMappingWithBazelDepsOnly(),
+              new HashMap<>());
+      for (ModuleExtensionUsage usage : module.getExtensionUsages()) {
+        try {
+          ModuleExtensionId moduleExtensionId =
+              ModuleExtensionId.create(
+                  labelConversionContext.convert(usage.getExtensionBzlFile()),
+                  usage.getExtensionName());
+          extensionUsagesTableBuilder.put(moduleExtensionId, module.getKey(), usage);
+        } catch (LabelSyntaxException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)